server/exercise/src/main/scala/com/eigengo/lift/exercise/UserExercisesClassifier.scala

```diff
@@ -79,18 +79,19 @@ class UserExercisesClassifier(sessionProps: SessionProperties, modelProps: Props
     // TODO: refactor code so that the following assumptions may be weakened further!
     case sdwls: ClassifyExerciseEvt =>
       require(
-        sdwls.sensorData.map(_.location).toSet == Sensor.sourceLocations && sdwls.sensorData.map(_.location).size == Sensor.sourceLocations.size,
-        "for each sensor location, there is a unique and corresponding member in the sensor data for the `ClassifyExerciseEvt` instance"
+        sdwls.sensorData.map(_.location).toSet == Sensor.sourceLocations && sdwls.sensorData.forall(_.data.nonEmpty),
+        "all sensor locations are present in the `ClassifyExerciseEvt` instance and have data"
       )
-      val sensorMap = sdwls.sensorData.groupBy(_.location).mapValues(_.flatMap(_.data))
-      val blockSize = sensorMap(SensorDataSourceLocationWrist).length
+      // (SensorDataSourceLocation, Int) -> List[SensorData]
+      val sensorMap: Map[SensorDataSourceLocation, List[List[SensorData]]] = sdwls.sensorData.groupBy(_.location).mapValues(_.map(_.data))
+      val blockSize = sensorMap(SensorDataSourceLocationWrist).head.length
       require(
-        sensorMap.values.forall(_.length == blockSize),
-        "all sensor data locations have a common data length"
+        sensorMap.values.forall(_.forall(_.length == blockSize)),
+        "all sensor data location points have a common data length"
```

Comment:
While this is a reasonable assumption now, the number of samples can differ in the future (e.g. switching the watch to a 50 Hz sampling rate).